index scan

bright-starry-sky opened the pull request.

darionyaphet commented on src/common/base/NebulaKeyUtils.cpp:
```diff
@@ -88,6 +90,73 @@ std::string NebulaKeyUtils::kvKey(PartitionID partId, const folly::StringPiece&
+    std::string key;
+    key.reserve(kIndexPrefix.size() + kEdgeIndexLen + raw.size());
+    key.append(reinterpret_cast<const char*>(&item), sizeof(int32_t))
+            .append(kIndexPrefix.data(), kIndexPrefix.size())
```
please alignment thanks

bright-starry-sky commented:
1: Using kIndex instead of kData. 
2: Removed kIndexPrefix from index key. 
3: Removed IndexItem from all thrift requests. 
4: Moved hint from storage.thrift to common.thrift. 
5:fix alignment problem.

bright-starry-sky commented:
1, Resolve conflict. 
2, Directory structure refactoring

bright-starry-sky commented:
Please hold this pr. A code synchronization is required with #1360

bright-starry-sky commented:
1: Resolve the conflict
2: Reconcile pr #1360
3: Disable the property data get when no columns need to be returned.

bright-starry-sky commented on src/interface/common.thrift:
```diff
@@ -100,4 +102,22 @@ struct Pair {
+struct IndexHintItem {
+    1: binary                first_str,
+    2: binary                second_str,
+    3: SupportedType         type,
```
It seems that `type` can be delete. let me check.

bright-starry-sky commented:
Remove `type` from `IndexHintItem`
Ready to review

bright-starry-sky commented:
1, Resolved conflict.
2,Improved error handling.

dangleptr commented:
Let's go on after #1360 merged.

dangleptr commented on src/common/base/NebulaKeyUtils.h:
```diff
@@ -276,6 +276,27 @@ class NebulaKeyUtils final {
     }
 
+    static VertexID getIndexVertexID(const folly::StringPiece& rawKey) {
+         auto offset = rawKey.size() - sizeof(VertexID);
```
You'd better check the key.size()

dangleptr commented on src/common/base/NebulaKeyUtils.h:
```diff
@@ -276,6 +276,27 @@ class NebulaKeyUtils final {
+     }
+
+    static VertexID getIndexSrcId(const folly::StringPiece& rawKey) {
+        auto offset = rawKey.size() -
```
ditto

dangleptr commented on src/common/base/NebulaKeyUtils.h:
```diff
@@ -276,6 +276,27 @@ class NebulaKeyUtils final {
+    }
+
+    static VertexID getIndexDstId(const folly::StringPiece& rawKey) {
+        auto offset = rawKey.size() - sizeof(VertexID);
```
ditto

dangleptr commented on src/common/base/NebulaKeyUtils.h:
```diff
@@ -276,6 +276,27 @@ class NebulaKeyUtils final {
+    }
+
+    static EdgeRanking getIndexRank(const folly::StringPiece& rawKey) {
+        auto offset = rawKey.size() - sizeof(VertexID) - sizeof(EdgeRanking);
```
ditto

dangleptr commented on src/interface/common.thrift:
```diff
@@ -103,9 +103,20 @@ struct Pair {
+    3: required list<ColumnDef>  cols,
+}
+
+struct IndexHintItem {
```
Could you add some comments for this structure.

dangleptr commented on src/interface/common.thrift:
```diff
@@ -103,9 +103,20 @@ struct Pair {
+    2: binary                second_str,
+}
+
+struct IndexHint {
```
ditto

dangleptr commented on src/interface/storage.thrift:
```diff
@@ -188,6 +208,13 @@ struct EdgePropRequest {
 
+struct IndexScanRequest {
+    1: common.GraphSpaceID       space_id,
+    2: list<common.PartitionID>  parts,
```
Do we really need this field?

dangleptr commented on src/interface/storage.thrift:
```diff
@@ -188,6 +208,13 @@ struct EdgePropRequest {
+    1: common.GraphSpaceID       space_id,
+    2: list<common.PartitionID>  parts,
+    3: common.IndexHint          hint,
+    4: list<string>              return_columns,
```
Each element is the column name?

bright-starry-sky replied:
Yes , each element is a column name. the column list corresponds to yield clause.

dangleptr commented on src/interface/storage.thrift:
```diff
@@ -412,4 +439,8 @@ service StorageService {
     GetUUIDResp getUUID(1: GetUUIDReq req);
+
+    // Interfaces for edge and vertex index scan
+    ScanVertexResponse scanVertexIndex(1: IndexScanRequest req);
```
LookUpVertexIndex

dangleptr commented on src/interface/storage.thrift:
```diff
@@ -412,4 +439,8 @@ service StorageService {
+
+    // Interfaces for edge and vertex index scan
+    ScanVertexResponse scanVertexIndex(1: IndexScanRequest req);
+    ScanEdgeResponse   scanEdgeIndex(1: IndexScanRequest req);
```
Ditto

dangleptr commented on src/interface/storage.thrift:
```diff
@@ -188,6 +208,13 @@ struct EdgePropRequest {
     5: list<PropDef> return_columns,
 }
 
+struct IndexScanRequest {
```
LookUpIndexRequest

dangleptr commented on src/interface/storage.thrift:
```diff
@@ -118,6 +126,18 @@ struct QueryResponse {
     4: optional list<VertexData> vertices,
 }
 
+struct ScanVertexResponse {
```
LookUpVertexIndexResp

dangleptr commented on src/interface/storage.thrift:
```diff
@@ -118,6 +126,18 @@ struct QueryResponse {
+    3: optional list<VertexIndexData>      rows,
+}
+
+struct ScanEdgeResponse {
```
LookUpEdgeIndexResp

dangleptr commented on src/storage/client/StorageClient.inl:
```diff
@@ -234,6 +234,87 @@ folly::Future<StatusOr<Response>> StorageClient::getResponse(
 
+
+template<class Request, class RemoteFunc, class Response>
+folly::SemiFuture<StorageRpcResponse<Response>> StorageClient::getResponse(
```
Why not reuse current 'getResponse' method ?

dangleptr commented on src/storage/index/ScanEdgeIndexProcessor.cpp:
```diff
@@ -0,0 +1,168 @@
+    auto indexId = hint_.index_id;
+    auto iRet = schemaMan_->getEdgeIndex(spaceId_, indexId);
+    if (!iRet.ok()) {
+        putResultCodes(cpp2::ErrorCode::E_LOAD_META_FAILED, req.get_parts());
```
Maybe INDEX_NOT_FOUND is a better choice

dangleptr commented on src/storage/index/ScanEdgeIndexProcessor.cpp:
```diff
@@ -0,0 +1,168 @@
+
+    std::vector<folly::Future<std::pair<PartitionID, kvstore::ResultCode>>> results;
+    for (auto& part : req.get_parts()) {
+        results.emplace_back(asyncProcess(part, indexId, edgeType));
```
I think search all parts one by one is enough here.

critical27 commented on src/storage/index/LookUpIndexBaseProcessor.inl:
```diff
@@ -0,0 +1,201 @@
+        if (hint_.hint_items.size() != 1) {
+            return cpp2::ErrorCode::E_INVALID_INDEX_HINT;
+        }
+        range_ = std::make_pair(hint_.hint_items[0].get_first_str(),
```
When we can only scan index within range [start, end], user specify it or graph?

critical27 commented on src/storage/index/LookUpIndexBaseProcessor.inl:
```diff
@@ -0,0 +1,201 @@
+    } else {
+        prefix_.reserve(128);
+        for (auto& hint : hint_.hint_items) {
+                prefix_.append(hint.get_first_str());
```
Who take care of the index hint order? I guess graph check the what kind of index we have, and if thery are the first N columns?

critical27 commented on src/storage/index/LookUpEdgeIndexProcessor.cpp:
```diff
@@ -0,0 +1,168 @@
+                                   begin,
+                                   end,
+                                   &iter);
+        if (ret == kvstore::SUCCEEDED) {
```
Similar to the code block when we use prefix, maybe we can unify them to base.

bright-starry-sky commented:
Please hold this pr review, I'm refactoring the code.
I'll push down the optimization of the index scan to the storage layer.

darionyaphet commented on src/interface/common.thrift:
```diff
@@ -103,9 +103,20 @@ struct Pair {
+}
+
+struct IndexHintItem {
+    1: binary                first_str,
```
what is the first and second string’s meaning ？

bright-starry-sky replied:
> what is the first and second string’s meaning ？

 Aha , I'm refactoring the index scan logic now. Most of the existing code will be discarded. 
By the way, the first and second string is a data scan range-pair,  I'll use the native expression instead of it in the storage layer.

bright-starry-sky replied:
> what is the first and second string’s meaning ？


what is the first and second string’s meaning ？

The index scan filter logic push down to storage layer.

dangleptr commented on src/storage/client/StorageClient.h:
```diff
@@ -222,7 +237,8 @@ class StorageClient {
         std::unordered_map<HostAddr, Request> requests,
-        RemoteFunc&& remoteFunc);
+        RemoteFunc&& remoteFunc,
+        GetIdFunc f);
```
GetPartIDFunc maybe a better name. 
And could we give it a default value?

dangleptr commented on src/storage/index/IndexPolicyMaker.h:
```diff
@@ -0,0 +1,117 @@
+    ACCURATE_SCAN      = 0x03,
+};
+
+using OperatorList = std::vector<std::pair<std::pair<std::string, VariantType>, bool>>;
```
Add comments for this structure.

dangleptr commented on src/storage/index/IndexPolicyMaker.h:
```diff
@@ -0,0 +1,117 @@
+    /**
+     * No index column match. seek scan and using expression condition filter.
+     */
+    SEEK_SCAN          = 0x01,
```
INDEX_SCAN maybe a better name.

dangleptr commented on src/storage/index/IndexPolicyMaker.h:
```diff
@@ -0,0 +1,117 @@
+     * Few index column match, index scan in subset.
+     * And using expression condition filter for remaining index column.s
+     */
+    PREFIX_SCAN        = 0x02,
```
PREFIX_SCAN and ACCURATE_SCAN are the same things.  Unify them.

bright-starry-sky replied:
They are different, `ACCURATE_SCAN` does not need to use expression filtering ,  PREFIX_SCAN and SEEK_SCAN needed. I will unify PREFIX_SCAN and SEEK_SCAN . 
WDYT?

dangleptr commented on src/storage/index/IndexPolicyMaker.h:
```diff
@@ -0,0 +1,117 @@
+namespace nebula {
+namespace storage {
+
+enum class PolicyType : uint8_t {
```
I am confused about this enum.  Why we need it.

bright-starry-sky replied:
When we traverse the expression tree, we terminate the traversal when  `OR` Operator occur.  if so , the  `PolicyType = SIMPLE_POLICY`
Also, when create scan policy, if the `PolicyType = SIMPLE_POLICY` ,  The logic of policy creation will be skipped。

bright-starry-sky replied:
resolved.

darionyaphet commented on src/storage/index/IndexExecutor.inl:
```diff
@@ -0,0 +1,329 @@
+                                                      cpp2::VertexIndexData* data) {
+    auto vId = NebulaKeyUtils::getIndexVertexID(key);
+    data->set_vertex_id(vId);
+    if (schema_ == nullptr) {
```
`schema_ == null` Is that possible?

bright-starry-sky replied:
> `schema_ == null` Is that possible?

Emm… the schema_ is a shared_ptr as below : 
std::shared_ptr<SchemaWriter>          schema_{nullptr};

darionyaphet replied:
when `checkReturnColumns` the schema have constructed.

BTW is the schema is null why return `SUCCEEDED`?

bright-starry-sky replied:
> when `checkReturnColumns` the schema have constructed.
> 
> BTW is the schema is null why return `SUCCEEDED`?

Aha, The result set such as `tag_index(VertexID , col1, col2...) `or `edge_index(srcVID, dstVID, Ranking, col1 , col2...)` , if the yield clause is empty , so we only need return (VertexID) or (srcVID, dstVID, Ranking) . So we don't have to handle with the schema, so return SUCCEEDED direct.

darionyaphet commented on src/storage/index/IndexExecutor.inl:
```diff
@@ -0,0 +1,329 @@
+    edge.set_ranking(rank);
+    edge.set_dst(dst);
+    data->set_key(edge);
+    if (schema_ == nullptr) {
```
similar with the above

dangleptr commented on src/storage/index/IndexPolicyMaker.cpp:
```diff
@@ -0,0 +1,149 @@
+                                   return col.get_name() == std::get<0>(tup);;
+                               });
+        if (it != operatorList_.end()) {
+            if (std::get<2>(*it) == RelationalExpression::Operator::EQ) {
```
We don't handle ">, <, >=, <=" ?

bright-starry-sky replied:
> We don't handle ">, <, >=, <=" ?

Yes , currently use expression to handle such operations  ">, <, >=, <=" .
 In fact, the best way is to handle with it in a subset of range.

dangleptr replied:
Yes,  full scan the whole index is too costly.

bright-starry-sky commented:
> Totally, the pr looks good to me now.
> Do you want to add more policies in the future?
> Especially for the case "A > 10" or "10 < A < 20"

Of course, in the future, we'll add different policies, for example ：
tag (col1 , col2 , col3) 
index (col1, col2)
where col1 > 10 and col1 < 20  
| where (col2 < 10 or col2 > 20) and col1 == 5
| col1 > 10 and col2 > 10
| Etc.

darionyaphet commented on src/storage/index/IndexPolicyMaker.cpp:
```diff
@@ -0,0 +1,149 @@
+            operatorList_.emplace_back(std::make_tuple(std::move(prop), std::move(v), rExpr->op()));
+            break;
+        }
+        case nebula::Expression::kFunctionCall : {
```
`kUUID` maybe same with `kFunctionCall `

bright-starry-sky replied:
Good point, In the future, we need add the logic when uuid supprot where clause in the future.

darionyaphet commented on src/storage/index/IndexPolicyMaker.cpp:
```diff
@@ -0,0 +1,149 @@
+            break;
+        }
+    }
+    if (hintNum == operatorList_.size() && !hasStr) {
```
why could not contain String ?

bright-starry-sky replied:
> why could not contain String ?

Because string is a variable-length field, We need to filter this field with expression. for example : 
index (c1, c2)
row1 ("ABC", "ABC")
row2("AB", "CABC")
where c1 == "ABC"    

the two rows has same prefix string "ABCABC" , we need using expression's eval() method to get the correct row row1.

darionyaphet replied:
Got~

darionyaphet commented on src/storage/index/IndexExecutor.h:
```diff
@@ -0,0 +1,102 @@
+    /**
+     * Details Scan index part as one by one.
+     **/
+    kvstore::ResultCode executeExecutionPlan(PartitionID part);
```
Sorry to say, the execution plan was representative the behavior of `Query Engine` do you mind use another name ?

bright-starry-sky replied:
> Sorry to say, the execution plan was representative the behavior of `Query Engine` do you mind use another name ?

Sounds reasonable，any Suggestions?

dangleptr replied:
I think its fine here.

dangleptr commented on src/storage/index/LookUpEdgeIndexProcessor.cpp:
```diff
@@ -0,0 +1,69 @@
+     */
+    auto ret = prepareRequest(req);
+    if (ret != cpp2::ErrorCode::SUCCEEDED) {
+        putResultCodes(ret, req.get_parts());
```
onFinished

darionyaphet replied:
When `putResultCodes` is called, onFinished() is included in this function.

dangleptr commented on src/storage/index/LookUpEdgeIndexProcessor.cpp:
```diff
@@ -0,0 +1,69 @@
+     */
+    ret = buildExecutionPlan(req.get_filter());
+    if (ret != cpp2::ErrorCode::SUCCEEDED) {
+        putResultCodes(ret, req.get_parts());
```
ditto

dangleptr commented on src/storage/index/LookUpVertexIndexProcessor.cpp:
```diff
@@ -0,0 +1,69 @@
+     */
+    auto ret = prepareRequest(req);
+    if (ret != cpp2::ErrorCode::SUCCEEDED) {
+        putResultCodes(ret, req.get_parts());
```
ditto

dangleptr commented on src/storage/index/LookUpVertexIndexProcessor.cpp:
```diff
@@ -0,0 +1,69 @@
+     */
+    ret = buildExecutionPlan(req.get_filter());
+    if (ret != cpp2::ErrorCode::SUCCEEDED) {
+        putResultCodes(ret, req.get_parts());
```
ditto

darionyaphet commented on src/storage/test/IndexScanTest.cpp:
```diff
@@ -0,0 +1,1487 @@
+/* Copyright (c) 2018 vesoft inc. All rights reserved.
```
2020

dangleptr commented on src/storage/index/IndexExecutor.inl:
```diff
@@ -0,0 +1,329 @@
+            iter->next();
+            continue;
+        }
+        ret = getDataRow(part, key);
```
We'd better find all valid edges/vertices firstly,  then get the properties from then.

laura-ding commented on src/storage/client/StorageClient.h:
```diff
@@ -306,6 +322,28 @@ class StorageClient {
+            if (!metaStatus.ok()) {
+                return metaStatus.status();
+            }
+            auto partMeta = metaStatus.value();
```
std::move(metaStatus).value()

laura-ding commented on src/storage/index/IndexExecutor.inl:
```diff
@@ -0,0 +1,329 @@
+                      schemaMan_->getEdgeSchema(spaceId_, tagOrEdge_) :
+                      schemaMan_->getTagSchema(spaceId_, tagOrEdge_);
+        if (!schema) {
+            VLOG(3) << "Can't find schema, spaceId "
```
why use `VLOG`

dangleptr commented on src/storage/index/IndexPolicyMaker.h:
```diff
@@ -0,0 +1,80 @@
+    std::string                              prefix_;
+    std::shared_ptr<nebula::cpp2::IndexItem> index_{nullptr};
+    bool                                     optimizedPolicy_{true};
+    bool                                     isIndexScan_{true};
```
maybe use a better name here.  For example,  skipFilter_ ?

laura-ding commented on src/storage/index/IndexExecutor.inl:
```diff
@@ -0,0 +1,329 @@
+            VLOG(3) << "Insert cache for vId " << vId << ", tagId " << tagOrEdge_;
+        }
+    } else {
+        VLOG(3) << "Missed partId " << partId << ", vId " << vId << ", tagId " << tagOrEdge_;
```
Why error log use VLOG(3)

laura-ding commented on src/storage/index/IndexExecutor.inl:
```diff
@@ -0,0 +1,329 @@
+                                                const folly::StringPiece& prop) {
+    using nebula::cpp2::SupportedType;
+    auto type = indexCols_[prop.str()];
+    return NebulaKeyUtils::decodeVariant(getIndexVal(key, prop), type);
```
`decodeVariant ` return `VariantType`, but here return `OptVariantType`, I think `decodeVariant ` need handle some error.

laura-ding commented on src/storage/index/IndexExecutor.inl:
```diff
@@ -0,0 +1,329 @@
+        }
+        offset += len;
+    }
+    return key.subpiece(offset, len);
```
Need to check the` key.len` with `offset + len`, if not equal, need return error

bright-starry-sky replied:
These are never going to be equal.

bright-starry-sky commented on src/common/base/NebulaKeyUtils.h:
```diff
@@ -262,16 +262,30 @@ class NebulaKeyUtils final {
+             *   have a problem of precision overflow. current return value is -nan.
+             */
+            auto i = *reinterpret_cast<const int64_t*>(&v);
+            i = -(std::numeric_limits<int64_t >::max() + i);
```
Sharp eyes 👍

bright-starry-sky commented:
Addressed comments.

dangleptr commented on src/storage/index/IndexExecutor.inl:
```diff
@@ -0,0 +1,333 @@
+        /**
+         * Need to filter result with expression if is not accurate scan.
+         */
+        if (skipFilter_ && !conditionsCheck(key)) {
```
skipFilter_ || !conditionsCheck(key) ?

bright-starry-sky replied:
I think the best is (skipFilter_ && !conditionsCheck(key)) , because we can step out the if block direct when skipFilter == false .

dangleptr replied:
I am confused. "skipFilter_" means skip the whole expression logic.

bright-starry-sky replied:
Your are right, I've changed the variable name from `skipFilter_` to `requiredFilter_`.
